v4/unit_cache.go

```diff
@@ -0,0 +1,84 @@
+/* Copyright 2022 Freerware
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package work
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/uber-go/tally"
+)
+
+// UnitCache represents the cache that the work unit manipulates as a result
+// of entity registration.
+type UnitCache struct {
+	m sync.Map
+
+	scope tally.Scope
+}
+
+var (
+	// ErrUncachableEntity represents the error that is returned when an attempt
+	// to cache an entity with an unresolvable ID occurs.
+	ErrUncachableEntity = errors.New("unable to cache entity - does not implement supported interfaces")
+)
+
+// Delete removes an entity from the work unit cache.
+func (uc *UnitCache) delete(entity interface{}) {
+	t := TypeNameOf(entity)
+	if id, ok := id(entity); ok {
+		if entitiesByID, ok := uc.m.Load(t); ok {
+			if entityMap, ok := entitiesByID.(*sync.Map); ok {
+				entityMap.Delete(id)
+				uc.scope.Counter(cacheDelete).Inc(1)
+			}
+		}
+	}
+}
+
+// Store places the provided entity in the work unit cache.
+func (uc *UnitCache) store(entity interface{}) (err error) {
+	id, ok := id(entity)
+	if !ok {
+		err = ErrUncachableEntity
```

Comment:
in this particular case, i agree i could've one either way. i generally like getting the zero value for free to avoid `return nil` (for none error scenarios) throughout the code. more of a preference i suppose.